go/pkg/client/cas.go

```diff
@@ -900,43 +892,79 @@ func (wt *writerTracker) Close() error {
 	return nil
 }
 
-func (c *Client) readBlobStreamed(ctx context.Context, hash string, sizeBytes, offset, limit int64, w io.Writer) (int64, error) {
-	if sizeBytes == 0 {
-		// Do not download empty blobs.
-		return 0, nil
-	}
-	wt := newWriteTracker(w)
-	name, wc, done, err := c.maybeCompressReadBlob(hash, sizeBytes, wt)
-	if err != nil {
-		return 0, err
-	}
-	n, err := c.readStreamed(ctx, name, offset, limit, wc)
+type resettableWriter interface {
+	io.Writer
+	Reset() error
+}
+
+type fileWriter struct {
+	f *os.File
+}
+
+func (fw *fileWriter) Write(p []byte) (int, error) {
+	return fw.f.Write(p)
+}
+
+func (fw *fileWriter) Reset() error {
+	err := fw.f.Truncate(0)
 	if err != nil {
-		return n, err
-	}
-	if err = wc.Close(); err != nil {
-		return n, err
-	}
-	if err := <-wt.ready; err != nil {
-		return n, err
-	}
-	close(wt.ready)
-	sz := sizeBytes - offset
-	if limit > 0 && limit < sz {
-		sz = limit
-	}
-	if err := <-done; err != nil {
-		return n, fmt.Errorf("Failed to finalize writing downloaded data downstream: %v", err)
+		return err
 	}
-	close(done)
-	// We can't reliably calculate hash without the full blob
-	if offset == 0 && limit <= 0 && wt.dg.Hash != hash {
-		return n, fmt.Errorf("CAS fetch calculated hash %s != expected hash %s", wt.dg.Hash, hash)
+	_, err = fw.f.Seek(0, 0)
+	return err
+}
+
+type bufferWriter struct {
+	b *bytes.Buffer
+}
+
+func (bw *bufferWriter) Write(p []byte) (int, error) {
+	return bw.b.Write(p)
+}
+
+func (bw *bufferWriter) Reset() error {
+	bw.b.Reset()
+	return nil
+}
+
+var DigestMismatchError = errors.New("CAS fetch digest mismatch")
+
+func (c *Client) readBlobStreamed(ctx context.Context, d digest.Digest, offset, limit int64, w resettableWriter) (n int64, err error) {
+	if d.Size == 0 {
+		// Do not download empty blobs.
+		return 0, nil
 	}
-	if wt.dg.Size != sz {
-		return n, fmt.Errorf("CAS fetch read %d bytes but %d were expected", n, sz)
+	closure := func() error {
+		if err = w.Reset(); err != nil {
+			return err
+		}
+		wt := newWriteTracker(w)
+		name, wc, done, e := c.maybeCompressReadBlob(d, wt)
+		if e != nil {
+			return e
+		}
+		n, err = c.readStreamed(ctx, name, offset, limit, wc)
+		if err != nil {
+			return err
+		}
+		if err = wc.Close(); err != nil {
+			return err
+		}
+		if err = <-wt.ready; err != nil {
+			return err
+		}
+		close(wt.ready)
+		if err = <-done; err != nil {
+			return fmt.Errorf("Failed to finalize writing downloaded data downstream: %v", err)
+		}
+		close(done)
+		// We can't reliably calculate hash without the full blob
+		if offset == 0 && limit <= 0 && wt.dg != d {
```

Comment:
Not this cl's fault, but I think instead of `offset == 0 && limit <= 0`, we should check `d.SizeBytes == sz`. The limit being > 0 doesn't mean we couldn't potentially be attempting to pull the entire blob.

Reply:
Done.